File: nestjs/src/auth/auth.controller.ts
import {
  Controller,
  Post,
  Body,
  Res,
  Req,
  HttpCode,
  HttpStatus,
  ValidationPipe,
  UnauthorizedException,
  UseGuards,
} from '@nestjs/common';
import type { Response, Request } from 'express';
import { AuthService } from './auth.service';
import { RegisterDto, LoginDto, AuthResponseDto } from './dto/auth.dto';
import { ConfigService } from '@nestjs/config';
import { JwtAuthGuard } from './guards/jwt-auth.guard';
import { GetUser } from './decorators/get-user.decorator';

@Controller('api/auth')
export class AuthController {
  constructor(
    private readonly authService: AuthService,
    private readonly configService: ConfigService,
  ) {}

  @Post('register')
  async register(
    @Body(ValidationPipe) registerDto: RegisterDto,
    @Res({ passthrough: true }) response: Response,
  ): Promise<AuthResponseDto> {
    const { accessToken, refreshToken } =
      await this.authService.register(registerDto);

    this.setRefreshTokenCookie(response, refreshToken);

    return { accessToken };
  }

  @Post('login')
  @HttpCode(HttpStatus.OK)
  async login(
    @Body(ValidationPipe) loginDto: LoginDto,
    @Res({ passthrough: true }) response: Response,
  ): Promise<AuthResponseDto> {
    const { accessToken, refreshToken } =
      await this.authService.login(loginDto);

    this.setRefreshTokenCookie(response, refreshToken);

    return { accessToken };
  }

  @Post('refresh')
  @HttpCode(HttpStatus.OK)
  async refresh(
    @Req() request: Request,
    @Res({ passthrough: true }) response: Response,
  ): Promise<AuthResponseDto> {
    const oldRefreshToken = request.cookies['jid'] as string | undefined;

    if (!oldRefreshToken) {
      throw new UnauthorizedException('No token provided');
    }

    const { accessToken, refreshToken } =
      await this.authService.refresh(oldRefreshToken);

    this.setRefreshTokenCookie(response, refreshToken);

    return { accessToken };
  }

  @Post('logout')
  @UseGuards(JwtAuthGuard)
  @HttpCode(HttpStatus.OK)
  async logout(
    @Req() request: Request,
    @Res({ passthrough: true }) response: Response,
    @GetUser('id') userId: string,
  ): Promise<{ message: string }> {
    const refreshToken = request.cookies['jid'] as string | undefined;

    if (refreshToken) {
      await this.authService.logout(userId, refreshToken);
    }

    response.clearCookie('jid', { path: '/api/auth/refresh' });

    return { message: 'Logged out' };
  }

  private setRefreshTokenCookie(
    response: Response,
    refreshToken: string,
  ): void {
    const isProduction =
      this.configService.get<string>('NODE_ENV') === 'production';

    response.cookie('jid', refreshToken, {
      httpOnly: true,
      secure: isProduction,
      sameSite: 'lax',
      path: '/api/auth/refresh',
      maxAge: 1000 * 60 * 60 * 24 * 7, // 7 days
    });
  }
}

File: springboot/src/main/java/com/example/springboot/auth/AuthController.java
package com.example.springboot.auth;

import com.example.springboot.auth.dto.AuthResponse;
import com.example.springboot.auth.dto.LoginRequest;
import com.example.springboot.auth.dto.RegisterRequest;
import jakarta.servlet.http.HttpServletResponse;
import jakarta.validation.Valid;
import org.springframework.http.HttpHeaders;
import org.springframework.http.ResponseCookie;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.Authentication;
import org.springframework.web.bind.annotation.CookieValue;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;
import org.springframework.web.server.ResponseStatusException;
import org.springframework.http.HttpStatus;

@RestController
@RequestMapping("/api/auth")
public class AuthController {
  private final AuthService authService;

  public AuthController(AuthService authService) {
    this.authService = authService;
  }

  @PostMapping("/register")
  public ResponseEntity<AuthResponse> register(@Valid @RequestBody RegisterRequest request) {
    AuthService.TokenPair tokens = authService.register(request);
    return withRefreshCookie(new AuthResponse(tokens.accessToken()), tokens.refreshToken());
  }

  @PostMapping("/login")
  public ResponseEntity<AuthResponse> login(@Valid @RequestBody LoginRequest request) {
    AuthService.TokenPair tokens = authService.login(request);
    return withRefreshCookie(new AuthResponse(tokens.accessToken()), tokens.refreshToken());
  }

  @PostMapping("/refresh")
  public ResponseEntity<AuthResponse> refresh(@CookieValue(value = "jid", required = false) String refreshToken) {
    if (refreshToken == null || refreshToken.isBlank()) {
      throw new ResponseStatusException(HttpStatus.UNAUTHORIZED, "No token provided");
    }
    AuthService.TokenPair tokens = authService.refresh(refreshToken);
    return withRefreshCookie(new AuthResponse(tokens.accessToken()), tokens.refreshToken());
  }

  @PostMapping("/logout")
  public ResponseEntity<String> logout(
      Authentication authentication,
      @CookieValue(value = "jid", required = false) String refreshToken,
      HttpServletResponse response) {

    String userId = authentication != null ? (String) authentication.getPrincipal() : null;
    if (userId != null && refreshToken != null) {
      authService.logout(userId, refreshToken);
    }

    ResponseCookie cleared = ResponseCookie.from("jid", "")
        .httpOnly(true)
        .secure(false)
        .path("/api/auth/refresh")
        .maxAge(0)
        .sameSite("Lax")
        .build();

    response.addHeader(HttpHeaders.SET_COOKIE, cleared.toString());
    return ResponseEntity.ok("Logged out");
  }

  private ResponseEntity<AuthResponse> withRefreshCookie(AuthResponse body, String refreshToken) {
    ResponseCookie cookie = ResponseCookie.from("jid", refreshToken)
        .httpOnly(true)
        .secure(false)
        .path("/api/auth/refresh")
        .maxAge(60L * 60 * 24 * 7)
        .sameSite("Lax")
        .build();

    return ResponseEntity.ok()
        .header(HttpHeaders.SET_COOKIE, cookie.toString())
        .body(body);
  }
}
